File: day2/TS/src/day2.ts
import fs from 'fs';

export function part1(filename: string): number {
    const reports: number [][] = fs.readFileSync(filename,'utf-8')
        .trim()
        .split('\n')
        .map(line => line.split(' ').map(Number))

    let safeCount = 0

    for(const report of reports) {
        //Räknar ut skillandne mellan nivåerna brevid varandra.
        const diffs = report.slice(1).map((level,index) => level - report[index])
        //Kollar så att skillnaden är mellan 1 och 3.
        const validDiffs = diffs.every(diff => Math.abs(diff) >= 1 && Math.abs(diff) <= 3)
        // Kontrollerar om rapporten är strikt stigande eller fallande
        // level: aktuellt element i arrayen
        // index: positionen för det aktuella elementet
        // arr: hela arrayen
        // Första elementet (index === 0) är alltid "sant", så vi börjar jämföra från andra elementet
        const strictlyIncreasing = report.every((level,index,arr) => index === 0 || arr[index-1] < level);
        const strictlyDecreasing = report.every((level,index,arr) => index === 0 || arr[index-1] > level);

        if(validDiffs && (strictlyDecreasing || strictlyIncreasing)) safeCount++
    }
    return safeCount
}

File: day2/python/day2.py
def part1(filename): 
    safe_count = 0

    with open(filename,'r') as file:
        for line in file:
            #Gär raden til en lista av siffror.
            report = list(map(int,line.strip().split()))
            #Räknar ut diffen på dom som är brevid varandra 
            diffs = [report[i] - report[i-1] for i in range(1,len(report))]
            #Kollar att skillnaden är mellan 1 och 3.
            #Jämför även med föregående
            if not all(1 <= abs(diff) <= 3 for diff in diffs):
                continue
            #Kontrollera att rpporten är antingen strikt stigande eller strikt fallande.
            #Strikt stigande: Varje tal är större än det föregånde
            #Strikt fallande: Varje tal är mindre än det föregånde
            if not (all(report[i] > report[i-1] for i in range (1,len(report)))or all (report[i] < report[i-1] for i in range(1,len(report)))):
                continue
            
            safe_count += 1

    return safe_count

def part2(filename):
    